solr/core/src/java/org/apache/solr/handler/component/HttpShardHandlerFactory.java

```diff
@@ -449,9 +556,83 @@ private static boolean hasReplicaType(Object o, String preferred) {
     }
   }
 
+  private final ReplicaListTransformerFactory randomRltFactory = (String configSpec, SolrQueryRequest request,
+      ReplicaListTransformerFactory fallback) -> shufflingReplicaListTransformer;
+  private ReplicaListTransformerFactory stableRltFactory;
+  private ReplicaListTransformerFactory defaultRltFactory;
+
+  /**
+   * Private class responsible for applying pairwise sort based on inherent replica attributes,
+   * and subsequently reordering any equivalent replica sets according to behavior specified
+   * by the baseReplicaListTransformer.
+   */
+  private static final class TopLevelReplicaListTransformer implements ReplicaListTransformer {
+
+    private final NodePreferenceRulesComparator replicaComp;
+    private final ReplicaListTransformer baseReplicaListTransformer;
+
+    public TopLevelReplicaListTransformer(NodePreferenceRulesComparator replicaComp, ReplicaListTransformer baseReplicaListTransformer) {
+      this.replicaComp = replicaComp;
+      this.baseReplicaListTransformer = baseReplicaListTransformer;
+    }
+
+    @Override
+    public void transform(List<?> choices) {
```

Comment:
I may be very wrong here, but I think the output of this method would be the same (or at least the same in practice) if we first run the transformer on all the list (which essentially either shuffles it or uses the new affinity code to define the order) and then run the `choices.sort(replicaComp)`, since this last one is stable, and will maintain the order of equal elements.
If that's the case, this method could be much simpler, and we don't need to find the boundaries, etc, plus the transformer gets called only once instead of once per group.

Reply:
I had to think about this for a minute, when I wrote it, and again now :-) ...

I think it actually _would_ be different. The key is that the affinity code is always based on the same absolute order; it's only the _rotation_ that changes. And if we apply the rotation to the overall list before applying other preferences, the stability of the sort doesn't help us.

Take for example a situation with replicas [A, B, C, D, E]. For the sake of argument, let's say that B and C will end up being tied for "most preferred", according to the inherent preference comparators. The "affinity" transformation, applied at the top level, would sort the replicas: [A, B, C, D, E], then rotate them according to some explicitly specified or hash-derived int param. The problem is that there's only one in 5 possible top-level rotations would result in C being ranked first: [C, D, E, A, B]. So the resulting order would be deterministic, but wouldn't distribute load evenly.

Applying the affinity rotation "in post" does end up distributing load evenly, since you're always rotating mod the number of _equivalent_ replicas per group.

Reply:
Ah, right, because the affinity just sets the rotation. But does it have to be? What's the difference between doing a rotations vs a shuffle with a seed? like:
`Collections.shuffle(replicas, new Random(hashValue));`

Reply:
True, random-with-seed should work, and I'm open to that. One use case that would benefit from a list-rotation-based implementation (as opposed to random-with-seed) would be if you wanted to set up a tiered or grouped routing system.

For example, replication factor of 2, say you have two distinct types of users with different access patterns (that would result in different cache usage patterns). Rather than simply routing deterministically for a given user/request, you could choose one of exactly 2 routing params, bifurcating traffic depending on expected use. With list-rotation, you could specify `routingParam=0` or `routingParam=1` and get the desired behavior; random-with-seed might not work that way, depending on the exact seeds chosen.

Granted you could achieve this tiered/grouped routing in other ways, but with a list-rotation-based implementation it would be trivial. The cost, as you say, would be some additional complexity (code-wise, not performance-wise) in `HttpShardHandlerFactory`.

In fact, it's this type of use case that I had in mind when providing support for the ability to specify `dividend` directly (as opposed to always hashing the routing param, which would be similarly opaque and thus incompatible with a tiered/grouped routing strategy).

Reply:
You mean, you want to make sure those two users get different affinity, right? vs with shuffle-with-seed, they could end up in the same nodes?
In any way, I think this PR looks good, we can improve things too in the future if we want.

Reply:
>make sure those two users get different affinity, right? vs with shuffle-with-seed, they could end up in the same nodes?

Exactly, yes; that's the idea. Thanks!